File: Kalorie/Kalorie/Features/MealTypeSheet/MealTypeSheetViewModel.swift
//
//  MealTypeSheetViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 12.06.2024.
//

import Foundation

final class MealTypeSheetViewModel: ObservableObject {

    // MARK: - Properties

    @Published private(set) var state: LoadingState<Void> = .loaded
    @Published var mealTypes: [MealTypeDomain]
    @Published var newMealName = ""
    @Published var newMealStart = Date.now
    @Published var newMealEnd = Date.now
    @Published var isAddFormVisible = false
    @Published var alertItem: AlertItem?

    private let onMealTypesChanged: () -> Void
    private let createMealType: any CreateMealTypeUseCaseProtocol
    private let deleteMealType: any DeleteMealTypeUseCaseProtocol
    private let updateMealTypeTimes: any UpdateMealTypeTimesUseCaseProtocol

    // MARK: - Init

    init(
        mealTypes: [MealTypeDomain],
        onMealTypesChanged: @escaping () -> Void = {},
        createMealType: any CreateMealTypeUseCaseProtocol,
        deleteMealType: any DeleteMealTypeUseCaseProtocol,
        updateMealTypeTimes: any UpdateMealTypeTimesUseCaseProtocol
    ) {
        self.mealTypes = mealTypes
        self.onMealTypesChanged = onMealTypesChanged
        self.createMealType = createMealType
        self.deleteMealType = deleteMealType
        self.updateMealTypeTimes = updateMealTypeTimes
    }

    // MARK: - Functions

    @MainActor
    func onCreateMealType() async {
        state = .loading
        defer { state = .loaded }
        do {
            let newMeal = try await createMealType(
                name: newMealName,
                startTime: newMealStart,
                endTime: newMealEnd,
                existingMealTypes: mealTypes
            )
            mealTypes.append(newMeal)
            mealTypes.sort { $0.startTime < $1.startTime }
            isAddFormVisible = false
            newMealName = ""
            onMealTypesChanged()
        } catch CreateMealTypeError.emptyName {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorEmptyName)
        } catch CreateMealTypeError.duplicateName {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorDuplicateName)
        } catch CreateMealTypeError.timeConflict {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorTimeConflict)
        } catch CreateMealTypeError.durationTooShort {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorDurationTooShort)
        } catch {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorUnexpected)
        }
    }

    @MainActor
    func onDelete(at index: Int) async {
        guard mealTypes.count > 1 else {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorLastMealType)
            return
        }
        state = .loading
        defer { state = .loaded }
        let mealType = mealTypes[index]
        do {
            try await deleteMealType(mealType)
            mealTypes.removeAll { $0.id == mealType.id }
            onMealTypesChanged()
        } catch {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorDeleteError)
        }
    }

    @MainActor
    func onMove(from source: IndexSet, to destination: Int) {
        let originalTimes = mealTypes.map { (startTime: $0.startTime, endTime: $0.endTime) }
        mealTypes.move(fromOffsets: source, toOffset: destination)
        for index in mealTypes.indices {
            let (startTime, endTime) = originalTimes[index]
            mealTypes[index] = MealTypeDomain(
                id: mealTypes[index].id,
                name: mealTypes[index].name,
                startTime: startTime,
                endTime: endTime
            )
        }
    }

    @MainActor
    func onSaveReorder() async {
        state = .loading
        defer { state = .loaded }
        do {
            try await updateMealTypeTimes(mealTypes)
            onMealTypesChanged()
        } catch {
            alertItem = AlertItem(title: L10n.MealTypeSheet.errorUnexpected)
        }
    }

    func onShowAddForm() {
        guard let possibleStart = mealTypes.map({ $0.endTime }).max() else {
            newMealStart = Date.now
            newMealEnd = Date.now.withAddedMinutes(minutes: 30)
            isAddFormVisible = true
            return
        }
        newMealStart = possibleStart
        newMealEnd = possibleStart.withAddedMinutes(minutes: 30)
        isAddFormVisible = true
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/mealtypesheet/MealTypeSheetViewModel.kt
package antoni.kalorie.features.mealtypesheet

import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.usecases.CreateMealTypeUseCaseProtocol
import antoni.kalorie.core.usecases.DeleteMealTypeUseCaseProtocol
import antoni.kalorie.core.usecases.UpdateMealTypeTimesUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import antoni.kalorie.core.utils.minutesSinceMidnight
import antoni.kalorie.core.utils.withAddedMinutes
import antoni.kalorie.core.utils.zoned
import java.time.Instant
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class MealTypeSheetViewModel(
    mealTypes: List<MealTypeDomain>,
    private val onMealTypesChanged: () -> Unit = {},
    private val createMealType: CreateMealTypeUseCaseProtocol,
    private val deleteMealType: DeleteMealTypeUseCaseProtocol,
    private val updateMealTypeTimes: UpdateMealTypeTimesUseCaseProtocol,
) : ViewModel() {

    // MARK: - Properties

    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.loaded)
    val state: StateFlow<LoadingState<Unit>> = _state
    val mealTypes = MutableStateFlow(mealTypes)
    private val _hasPendingReorder = MutableStateFlow(false)
    val hasPendingReorder: StateFlow<Boolean> = _hasPendingReorder
    val newMealName = MutableStateFlow("")
    val newMealStart = MutableStateFlow(Instant.now())
    val newMealEnd = MutableStateFlow(Instant.now())
    val isAddFormVisible = MutableStateFlow(false)
    val isExportPushed = MutableStateFlow(false)
    val alertItem = MutableStateFlow<AlertItem?>(null)

    // MARK: - Functions

    suspend fun onCreateMealType() {
        _state.value = LoadingState.Loading
        try {
            val newMeal = createMealType(
                name = newMealName.value,
                startMinutes = newMealStart.value.minutesSinceMidnight(),
                endMinutes = newMealEnd.value.minutesSinceMidnight(),
                existingMealTypes = mealTypes.value,
            )
            mealTypes.value = (mealTypes.value + newMeal).sortedBy { it.startMinutes }
            isAddFormVisible.value = false
            newMealName.value = ""
            onMealTypesChanged()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            alertItem.value = when (error) {
                is CreateMealTypeError.EmptyName -> AlertItem(titleRes = R.string.mealTypeSheet_error_emptyName)
                is CreateMealTypeError.DuplicateName -> AlertItem(titleRes = R.string.mealTypeSheet_error_duplicateName)
                is CreateMealTypeError.TimeConflict -> AlertItem(titleRes = R.string.mealTypeSheet_error_timeConflict)
                is CreateMealTypeError.DurationTooShort -> AlertItem(titleRes = R.string.mealTypeSheet_error_durationTooShort)
                else -> {
                    Log.error(error, Constants.LogCategory.MEAL_TYPE_SHEET)
                    AlertItem(titleRes = R.string.mealTypeSheet_error_unexpected)
                }
            }
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    suspend fun onDelete(index: Int) {
        if (mealTypes.value.size <= 1) {
            alertItem.value = AlertItem(titleRes = R.string.mealTypeSheet_error_lastMealType)
            return
        }
        _state.value = LoadingState.Loading
        val mealType = mealTypes.value[index]
        try {
            deleteMealType(mealType)
            mealTypes.value = mealTypes.value.filter { it.id != mealType.id }
            onMealTypesChanged()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MEAL_TYPE_SHEET)
            alertItem.value = AlertItem(titleRes = R.string.mealTypeSheet_error_deleteError)
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    fun onMove(from: Int, to: Int) {
        val originalTimes = mealTypes.value.map { it.startMinutes to it.endMinutes }
        val moved = mealTypes.value.toMutableList()
        val movedMealType = moved.removeAt(from)
        moved.add(if (to > from) to - 1 else to, movedMealType)
        mealTypes.value = moved.mapIndexed { index, mealType ->
            val (startMinutes, endMinutes) = originalTimes[index]
            MealTypeDomain(id = mealType.id, name = mealType.name, startMinutes = startMinutes, endMinutes = endMinutes)
        }
        _hasPendingReorder.value = true
    }

    suspend fun onSaveReorder() {
        if (!_hasPendingReorder.value) return
        _state.value = LoadingState.Loading
        _hasPendingReorder.value = false
        try {
            updateMealTypeTimes(mealTypes.value)
            onMealTypesChanged()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MEAL_TYPE_SHEET)
            alertItem.value = AlertItem(titleRes = R.string.mealTypeSheet_error_unexpected)
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    fun onShowAddForm() {
        val latestEndMinutes = mealTypes.value.maxOfOrNull { it.endMinutes }
        if (latestEndMinutes == null) {
            newMealStart.value = Instant.now()
            newMealEnd.value = Instant.now().withAddedMinutes(30.0)
            isAddFormVisible.value = true
            return
        }
        val now = Instant.now().zoned()
        val possibleStart = now.toLocalDate().atStartOfDay().plusMinutes(latestEndMinutes.toLong()).atZone(now.zone).toInstant()
        newMealStart.value = possibleStart
        newMealEnd.value = possibleStart.withAddedMinutes(30.0)
        isAddFormVisible.value = true
    }
}
